USQOneElectronOperator

xdvriend opened the pull request:
**Short description**
Adding the Unrestricted Second Quantized operators, in order to later perform unrestricted calculations.

**Related issues**
This PR aims to solve the first part of #558 

**Additional context**
Trying to follow the structure of the restricted SQOperators as much as possible.

xdvriend commented:
I've written some tests for the USQOneElectronOperator and they all pass. I have opened it for review. Let me know what you think @lelemmen , any feedback is welcome!

lelemmen commented on gqcp/include/Operator/SecondQuantized/USQOneElectronOperator.hpp:
```diff
@@ -0,0 +1,445 @@
+        }
+
+        return Eigen::Map<Eigen::Matrix<Scalar, Components, 1>>(expectation_values_alpha.data()), Eigen::Map<Eigen::Matrix<Scalar, Components, 1>>(expectation_values_beta.data());  // convert std::array to Vector
+    }
```
This code does not do what is expected. The expectation value of any scalar operator (be it restricted, or unrestricted, or generalized, ...) is a number. (A vector operator, analogously, yields a vector).

So, the alpha- and beta contributions should be summed together, which is not what happens here.

xdvriend replied:
This is fixed in the latest commits.

lelemmen commented on gqcp/include/Operator/SecondQuantized/USQOneElectronOperator.hpp:
```diff
@@ -0,0 +1,445 @@
+
+
+    /**
+     *  @return the dimension of the total generalized coefficient matrix
```
```suggestion
     *  @return the sum of the alpha and beta dimensions
```

Since we've gone through great lengths decoupling coefficients and matrix elements, I feel it's better to avoid this relation in the documentation.

lelemmen commented on gqcp/include/Operator/SecondQuantized/USQOneElectronOperator.hpp:
```diff
@@ -0,0 +1,445 @@
+    //     }
+
+    //     return result_alpha, result_beta;
+    // }
```
Could you remove this from the code and create an issue discussing that this method may be incorrectly implemented for `SQOneElectronOperator`?

xdvriend replied:
The commented code has been removed in the latest commits.

lelemmen commented on gqcp/include/Operator/SecondQuantized/USQOneElectronOperator.hpp:
```diff
@@ -0,0 +1,445 @@
+
+
+
+} // namespace GQCP
```
```suggestion
} // namespace GQCP

```

lelemmen commented on gqcp/include/Utilities/typedefs.hpp:
```diff
@@ -48,5 +48,8 @@ using sum_t = decltype(std::declval<T>() + std::declval<U>());
 
+template<typename T, typename U>
+using difference_t = decltype(std::declval<T>() - std::declval<U>());
+
```
I like that you've made the abstraction to `difference_t`!

However, the difference type is always the same as the sum type, and we already have `sum_t`. 😉

xdvriend replied:
This is fixed in the latest commits.

lelemmen commented on gqcp/include/Operator/SecondQuantized/USQOneElectronOperator.hpp:
```diff
@@ -0,0 +1,445 @@
+}
+
+
+
```
```suggestion


```

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -0,0 +1,191 @@
+    BOOST_CHECK_THROW(GQCP::ScalarUSQOneElectronOperator<double> O_3 (matrix, square_matrix), std::invalid_argument);
+    BOOST_CHECK_THROW(GQCP::ScalarUSQOneElectronOperator<double> O_4 (square_matrix, matrix), std::invalid_argument);
+}
+
```
```suggestion


```

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -0,0 +1,191 @@
+    // Initialize the reference and check the result
+    GQCP::QCMatrix<double> M_diff_ref (dim);
+    M_diff_ref <<  4.0,  4.0,
+                  4.0,  4.0;
```
Could you align these entries?

xdvriend replied:
this is fixed in the latest commits.

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -0,0 +1,191 @@
+    BOOST_CHECK_THROW(h.calculateExpectationValue(D_valid, D_invalid), std::invalid_argument);
+
+    BOOST_CHECK_NO_THROW(h.calculateExpectationValue(D_valid, D_valid));
+}
```
```suggestion
}

```

xdvriend commented:
Thanks for the feedback! I'll start making the requested changes right away.

xdvriend commented:
I reopened this PR for review. I changed/added all your requests and I added the missing tests to the regular SQOneElectronOperator as well. But there is always the possibility that some other minor errors crept in. So once again, any feedback is welcome!

lelemmen commented on gqcp/tests/Operator/SecondQuantized/SQOneElectronOperator_test.cpp:
```diff
@@ -238,3 +285,112 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+          3.0, 4.0;
+    const GQCP::ScalarSQOneElectronOperator<double> op (M1);
+
+    // initialize an alpha and beta density matrix, each one is chosen as a hermitian matrix.
```
```suggestion
    // Initialize an alpha and beta density matrix, each one is chosen as a Hermitian matrix.
```

lelemmen commented on gqcp/tests/Operator/SecondQuantized/SQOneElectronOperator_test.cpp:
```diff
@@ -238,3 +285,112 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+    
+    op.rotate(U);
+    BOOST_CHECK(op.parameters().isApprox(M1, 1.0e-08));
+
```
```suggestion
```

lelemmen commented on gqcp/tests/Operator/SecondQuantized/SQOneElectronOperator_test.cpp:
```diff
@@ -238,3 +285,112 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+     M1 << 1.0, 2.0, 3.0, 4.0,
+          5.0, 6.0, 7.0, 8.0,
+          9.0, 10.0, 11.0, 12,
+          13.0, 14.0, 15.0, 16.0;
```
Could you align these values on the decimal point?

lelemmen commented on gqcp/tests/Operator/SecondQuantized/SQOneElectronOperator_test.cpp:
```diff
@@ -238,3 +285,112 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+    GQCP::ScalarSQOneElectronOperator<double> op (M1);
+
+    // Initialize a transformation matrix
+    GQCP::JacobiRotationParameters J (2, 1, M_PI_2);
```
The use of macros is generally avoided. It's better to use boost's library to create the mathematical constant pi/2.

lelemmen commented on gqcp/tests/Operator/SecondQuantized/SQOneElectronOperator_test.cpp:
```diff
@@ -238,3 +285,112 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+    ref <<  1.0, 3.0, -2.0, 4.0,
+            9.0, 11.0, -10.0, 12.0,
+            -5.0, -7.0, 6.0, -8.0,
+            13.0, 15.0, -14.0, 16.0;
```
Could you align these values on the decimal point?

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -18,10 +18,11 @@
 #define BOOST_TEST_MODULE "USQOneElectronOperator"
 
 #include <boost/test/unit_test.hpp>
+#include <math.h>
```
In tests, there's a different order of includes:
1. The necessary boost test includes.
2. GQCP includes
3. Third-party includes (Eigen/...)
4. STL

So, if you're using `<math.h>` inside this test file, it should go at the end.

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -188,4 +190,119 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+    // initialize an alpha and beta density matrix, each one is chosen as a hermitian matrix.
+    GQCP::QCMatrix<double> d_alpha (dim);
+    d_alpha << 0.0, 1.0,
+               1.0, 0.0;
```
```suggestion
    GQCP::QCMatrix<double> D_alpha (dim);
    D_alpha << 0.0, 1.0,
               1.0, 0.0;
```

The 1-DM is usually given the symbol `D` instead of `d`.

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -188,4 +190,119 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+
+    GQCP::QCMatrix<double> d_beta (dim);
+    d_beta << 1.0, 0.0,
+              0.0, -1.0;
```
```suggestion
    GQCP::QCMatrix<double> D_beta (dim);
    D_beta << 1.0,  0.0,
              0.0, -1.0;
```

The 1-DM is usually given the symbol `D` instead of `d`.

lelemmen commented on gqcp/tests/Operator/SecondQuantized/SQOneElectronOperator_test.cpp:
```diff
@@ -238,3 +285,112 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+    ref << 5.0;
+
+    const auto ex_value = op.calculateExpectationValue(d);
+    BOOST_CHECK(ex_value.isApprox(ref, 1.0e-08));
```
```suggestion
    GQCP::QCMatrix<double> D (dim);
    D << 0.0, 1.0,
         1.0, 0.0;
    
    // Initialize a reference value and check the result.
    const double reference_expectation_value = 5.0;

    const auto expectation_value = op.calculateExpectationValue(D)(0);  // extract the 'scalar' from a one-dimensional vector
    BOOST_CHECK(expectation_value.isApprox(reference_expectation_value, 1.0e-08));
```

The 1-DM is usually given the symbol `D` instead of `d`.

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -188,4 +190,119 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+    ref << 2.0;
+
+    const auto ex_value = op.calculateExpectationValue(d_alpha, d_beta);
+    BOOST_CHECK(ex_value.isApprox(ref, 1.0e-08));
```
```suggestion
    const double ref_expectation_value = 2.0;

    const auto expectation_value = op.calculateExpectationValue(D_alpha, D_beta)(0);  // extract the scalar out of a one-dimensional vector
    BOOST_CHECK(expectation_value.isApprox(ref_expectation_value, 1.0e-08));
```

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -188,4 +190,119 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+    op.rotate(U);
+    BOOST_CHECK(op.alphaParameters().isApprox(M1, 1.0e-08));
+    BOOST_CHECK(op.betaParameters().isApprox(M1, 1.0e-08));
+
```
```suggestion
```

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -188,4 +190,119 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+
+
+/**
+ * Check whether or not the rotate with transformation matrix method works as expected
```
```suggestion
 * Check whether or not the transformation with a transformation matrix method works as expected
```

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -188,4 +190,119 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+
+
+/**
+ * Check whether or not the rotate with transformation matrix method works as expected
```
```suggestion
 * Check whether or not the rotate with a unitary transformation matrix method works as expected
```

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -188,4 +190,119 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+
+
+/**
+ * Check whether or not the rotate with transformation matrix method works as expected
```
```suggestion
 * Check whether or not the rotate with Jacobi rotation parameters method works as expected
```

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -188,4 +190,119 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+    M1 << 1.0, 2.0, 3.0, 4.0,
+          5.0, 6.0, 7.0, 8.0,
+          9.0, 10.0, 11.0, 12,
+          13.0, 14.0, 15.0, 16.0;
```
Could you align these values on the decimal point?

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -188,4 +190,119 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+    GQCP::ScalarUSQOneElectronOperator<double> op (M1, M1);
+
+    // Initialize a transformation matrix
+    GQCP::JacobiRotationParameters J (2, 1, M_PI_2);
```
I have the same remark here: it's better to use boost for mathematical constants rather than a preprocessor macro.

lelemmen commented on gqcp/tests/Operator/SecondQuantized/USQOneElectronOperator_test.cpp:
```diff
@@ -188,4 +190,119 @@ BOOST_AUTO_TEST_CASE ( calculateExpectationValue_throw ) {
+    ref <<  1.0, 3.0, -2.0, 4.0,
+            9.0, 11.0, -10.0, 12.0,
+            -5.0, -7.0, 6.0, -8.0,
+            13.0, 15.0, -14.0, 16.0;
```
Could you align these values on the decimal point?

lelemmen commented:
@xdvriend Nice work! I'd like to give you some more advice as well. You can use `git commit --amend` to add the recently staged files to the previous commit. This would avoid duplicate commit messages, like 'fixed suggestions from lelemmen'.

xdvriend commented:
Thanks! And I'll be sure to keep that in mind for the next PR.
